### ISSUE TITLE
Implement Auto-Swap Service for STRK Transfers

### ISSUE THREAD

**Issue by Abeeujah** [BlockheaderWeb3-Community/autoswappr-backend#1]
## Overview

This service processes transfer events emitted by the STRK token contract. Its primary role is to automate token swaps for users who have subscribed to the auto-swap feature on our platform.

## Payload

The service should listen for incoming payloads containing the following fields:
- `from` : Sender's wallet address.
- `to` : Receiver's wallet address.
- `value` : Amount of STRK tokens transferred.

## Functionality

1. **Check Subscription:**
    - Query the `swap_subscription` table to verify if the `to` address has subscribed for auto-swap.
    - If subscribed, retrieve the `to_token` (stable token) they wish to swap to.

2. **Validate Swap Preferences:**
    - Query the `swap_subscription_from_token` table to confirm:
        - They subscribed to auto-swap from the STRK token.
        - The percentage of the received `value` they want to swap.

3. **Perform Swap:**
    - Use the following information to make a smart contract call to the EKUBO contract:
        - `from_token`: STRK.
        - `to_token`: The stable token retrieved.
        - `amount`: Calculated from the `value` and the specified percentage.

**Acceptance Criteria**
- Service should validate payload integrity before processing.
- Only subscribed users should trigger a swap operation.
- Accurate smart contract calls must be made to EKUBO with relevant details.
- Errors (e.g., invalid subscriptions or failed contract calls) should be logged for debugging.

### Examples

**Input Payload:**
```json
{
  "from": "0xSenderAddress",
  "to": "0xReceiverAddress",
  "value": 1000
}
```

**Database Queries:**

- `swap_subscription` for `wallet_address` and `to_token` (e.g., USDT).
- `swap_subscription_from_token` for `from_token` (STRK) and `percentage` (e.g., 50%).

**Expected Smart Contract Call:** Swap 50% of 1000 STRK (500 STRK) to USDT via EKUBO.

#### Dependencies
- STRK token contract.
- EKUBO smart contract.
- Database tables: swap_subscription and swap_subscription_from_token.

#### Notes
- Ensure security checks for the smart contract call.
- Plan for retries or fallback mechanisms in case of contract call failures.

**Comment by 0xdevcollins** [comment]
Is this issue still available?

My Background
I am a Full Stack Blockchain Developer with proficiency in Solidity/Rust, Next.js, TypeScript, React, and Node.js. I've demonstrated my ability to adjust to a variety of requirements, perform well under pressure, and reliably produce user-centric blockchain solutions with 46 significant contributions spread over 12 OnlyDust projects.

**Comment by mimisavage** [comment]
Can I work on this, please?

**Comment by GideonBature** [comment]
Could I try solving this?

my name is Gideon Bature, a Software Engineer with experience working on Rust projects. If assigned this task, will send my first PR within 48 hours.

Thank you!

**Comment by chigozzdevv** [comment]
Could I take over this issue? Thanks :)

**Comment by Birdmannn** [comment]
I can handle this, chief

**Comment by CEOliam** [comment]
May I try my hand at this?

**Comment by Michaelkingsdev** [comment]
I'd love to work on this!

**Comment by Benjtalkshow** [comment]
May I take this issue on?

**Comment by obacollins-lab** [comment]
Hi @GideonBature There's a database model if I'm correct, what you meant to ask is how to represent your request and response; if correct, you'd want to make use of structs and the `serde` crate for serializing and deserializing. an example might look like this:
```rust
#[derive(Serialize, Deserialize)]
struct Request {
  from: String,
  to: String,
  value: i32
}
```

**Comment by Abeeujah** [comment]
Hi @GideonBature There's a database model if I'm correct, what you meant to ask is how to represent your request and response; if correct, you'd want to make use of structs and the serde crate for serializing and deserializing. an example might look like this:

```rust
#[derive(Serialize, Deserialize)]
struct Request {
  from: String,
  to: String,
  value: i32
}
```

**Comment by GideonBature** [comment]
Yes, I meant apart from the migration file where the table schemas are defined, I didn't see any models in the codebase. 

Also, Please do you have any sample of how making a smart contract call to EKUBO? Have checked their docs and can't seems to find anything.

**Comment by sprtd** [comment]
> Yes, I meant apart from the migration file where the table schemas are defined, I didn't see any models in the codebase.
> 
> Also, Please do you have any sample of how making a smart contract call to EKUBO? Have checked their docs and can't seems to find anything.

@GideonBature For further queries, please feel free to ask in the [telegram](https://t.me/+PJfVLZUFylFjZDhk)

**Comment by GideonBature** [comment]
> > Yes, I meant apart from the migration file where the table schemas are defined, I didn't see any models in the codebase.
> > Also, Please do you have any sample of how making a smart contract call to EKUBO? Have checked their docs and can't seems to find anything.
> 
> @GideonBature For further queries, please feel free to ask in the [telegram](https://t.me/+PJfVLZUFylFjZDhk)

Alright, no problem. Thank you!

**Comment by sprtd** [comment]
@GideonBature any progress with this

**Comment by GideonBature** [comment]
Not yet...

```Accurate smart contract calls must be made to EKUBO with relevant details.```

Still haven't gotten an idea on how to implement this.

**Comment by Jagadeeshftw** [comment]
I would like to work on this @sprtd . I have good knowledge in rust

**Comment by 1nonlypiece** [comment]
Can I take this @sprtd ?

**Comment by Jagadeeshftw** [comment]
Thanks for assigning it to me @sprtd . I will start work on it immediately
